### ISSUE TITLE
Struct with a single float3 appears to be padded to the size of a float4, even with std430 layout.

### ISSUE THREAD

**Issue by BlenderSleuth** [shader-slang/slang#5806]
Hi there,

I have a struct with a single float3 member, which I use so I can add some member functions.
```
struct MaterialProperty {
    float3 value;

    float3 get() {
        return value;
    }
}
```

When I add this as a member of another struct, it appears to have the size of a float4 because the float afterwards has an offset of 16. This struct:
```
struct MaterialData {
    MaterialProperty base_colour;
    float roughness;
    MaterialProperty normal;
    uint flags;
}
StructuredBuffer<MaterialData> materials;
```

Gets compiled with the following offsets (from the RenderDoc dissassembly):
```
struct MaterialProperty_std430 {
  float3 value : [[Offset(0)]];
}
struct MaterialData_std430 {
  MaterialProperty_std430 base_colour : [[Offset(0)]];
  float roughness : [[Offset(16)]];
  MaterialProperty_std430 normal : [[Offset(32)]];
  uint flags : [[Offset(48)]];
}
```
An "unpacking" function seems to be defined that does the opposite of its name and packs the larger `_std430` version of the struct into the normal version with the correct offsets:
```
struct MaterialProperty {
  float3 value : [[Offset(0)]];
}
struct MaterialData {
  MaterialProperty base_colour : [[Offset(0)]];
  float roughness : [[Offset(12)]];
  MaterialProperty normal : [[Offset(16)]];
  uint flags : [[Offset(28)]];
}

MaterialData unpackStorage@2(MaterialData_std430 _138) {
  MaterialProperty_std430 _140 = _138.base_colour;
  MaterialProperty _141 = unpackStorage(_140);
  float _149 = _138.roughness;
  MaterialProperty_std430 _150 = _138.normal;
  MaterialProperty _151 = unpackStorage(_150);
  uint _152 = _138.flags;
  MaterialData _153 = CompositeConstruct({_141, _149, _151, _152});
  return _153;
}

MaterialProperty unpackStorage@3(MaterialProperty_std430 _144) {
  float3 _146 = _144.value;
  MaterialProperty _147 = CompositeConstruct({_146});
  return _147;
}
```

Is this the intended behaviour of the std430 layout? From the OpenGL 4.5 spec section 7.6.2.2 (Standard Uniform Block Layout), rule 9 of the std140 layout spec defines structures to have padding on the end to round up to a vec4. This property is then explicitly voided by the description of std430 ("that the base alignment and stride ... of structures in rule 9 are not rounded up a multiple of the base alignment of a vec4." In this case, the MaterialProperty struct shouldn't pad to 16 bytes under std430, if I'm understanding correctly. 

The issue appears to be caused by defining a nested struct with the float3 member; using a non-nested float3 instead has the correct offsets. I'm aware the scalar layout option avoids a lot of these issues, but ideally I'd like to keep the float3 16-byte alignment guarantee of std430. I can also pack things into a float4, but of course that isn't as ergonomic.

**Comment by csyonghe** [comment]
Yes, I believe what slang generates is the correct std430 behavior. See GLSL example here: https://godbolt.org/z/jvn6eE4of

**Comment by csyonghe** [comment]
From the OpenGL spec:
<img width="607" alt="image" src="https://github.com/user-attachments/assets/c7fc5e22-5c7d-418e-83e0-d626ec17149f">

- According to rule (3), `float3` will have alignment = 16, this is *not* invalidated by std430 rule.
- According to rule (9), `MaterialProperty` should have alignment = 16 because `MaterialProperty::value` has alignment = 16.
- According to rule (9) when laying out `MaterialData`, offset of `MaterialData::roughness` is "the base offset of the member following the structure `MaterialData::base_colour`",  thus "is rounded up to the next multiple of the base alligment of the structure (16)", so `MaterialData::base_colour` starts at offset `alignUp(12, 16) == 16`.

**Comment by csyonghe** [comment]
I am going to close this issue as answered, let us know if you still have questions regarding std430 rules.

**Comment by BlenderSleuth** [comment]
Thanks very much for looking into it for me, that makes sense. I think the missing piece was that custom structs cannot have a size less than their alignment (à la C structs), and that the built-in float3 is simply special in this regard.
